File: Java/JavaUtils/src/javautils/message/IMessageSerializer.java
package javautils.message;

import java.io.DataInputStream;
import java.io.DataOutputStream;

/*******************************************************************************
 * Generic message serializer interface.
 * 
 * @param <E>
 ******************************************************************************/
public interface IMessageSerializer<E>
{
    /***************************************************************************
     * Writes this class to the supplied stream.
     * 
     * @param stream
     * @throws Exception
     **************************************************************************/
    public void toBinaryStream( DataOutputStream stream ) throws Exception;

    /***************************************************************************
     * Reads and constructs a new object of type E from the supplied stream.
     * 
     * @param stream
     * @return
     * @throws Exception
     **************************************************************************/
    public E fromBinaryStream( DataInputStream stream ) throws Exception;
}

File: C#/ChatterBox/ChatterBox/IMessageSerializer.cs
using System.IO;

namespace ChatterBox
{
    interface IMessageSerializer<T>
    {
        /// <summary>
        /// Writes a message to the supplied BinaryWriter.
        /// </summary>
        /// <param name="writer"></param>
        void toBinaryStream( BinaryWriter writer );

        /// <summary>
        /// Reads a message from the supplied BinaryReader.
        /// </summary>
        /// <param name="reader"></param>
        /// <returns></returns>
        T fromBinaryStream( BinaryReader reader );
    }
}
